src/pip/_internal/resolution/resolvelib/provider.py

```diff
@@ -115,11 +115,18 @@ def get_preference(
         self,
         resolution,  # type: Optional[Candidate]
         candidates,  # type: Sequence[Candidate]
-        information  # type: Sequence[Tuple[Requirement, Candidate]]
+        information  # type: Sequence[Tuple[Requirement, Optional[Candidate]]]
     ):
         # type: (...) -> Any
-        # Use the "usual" value for now
-        return len(candidates)
+        """Return a sort key to determine what dependency to look next.
+
+        A smaller value makes a dependency higher priority. We put direct
+        (user-requested) dependencies first since they may contain useful
+        user-specified version ranges. Users tend to expect we to catch
+        problems in them early as well.
```

Comment:
```suggestion
        A smaller value makes a dependency higher priority. We put direct
        (user-requested) dependencies first since they may contain useful
        user-specified version ranges. Users tend to expect us to catch
        problems in them early as well.
```